### PR TITLE
Fix(Parser): Update parser to deal with preamble better

### LINKED ISSUE #166: Issue: Negative indices in issue commands cause a different error message from out of range indices

### LINKED ISSUE #167: Resident: Negative indices in issue commands cause a different resident message from out of range indices

### LINKED ISSUE #199: [PE-D] rdel 0 and rdel 7 have different error messages
As in title, rdel 0 and rdel 7 have different error messages. 
rdel 0: invalid command format
rdel 7: the resident index provided is invalid

-------------
Labels: `severity.Low` `type.FunctionalityBug`
original: nickyfoo/ped#1

### LINKED ISSUE #293: Bug(Room): Keying in empty oedit or odel does not show command description

### LINKED ISSUE #388: Suggestion: Edit commands should specify that a field is required
## Current behavior
Edit commands (`redit`, `oedit`, `iedit`) usage summary (in-app) do not specify that at least one optional field is required.

## Suggestion
They should specify that restriction.

## Justification
It is part of the restrictions of those commands and should appear in the usage summary.

### PR DESCRIPTION
**Pull request by DrWala**
### What this does:
- Fixes #166 
- Fixes #167 
- Fixes #199
- Fixes #293 

The parser now deals with non-integer and integer input separately. This
allows for better error messages to be raised to the user to explain why
the entered preamble used in index values may be failing.

This PR also separates non-integer input to throw an
illegalargumentexception instead of a parseexception, which is now
reserved for integer input that fails (<=0, > range of list, etc).

Additionally, this also fixes #388. This fix is done here as the master branch does not allow this specific fix to be tested properly.

### How to test
Run index-based commands (edit, delete) and ensure the behaviour is as follows:
1. Command without parameters should show command usage [ `oedit` ] should show "invalid command format [how to use]"
2. Command without non-integer index values should show command usage [ `oedit asdfa`, `oedit  1 asdfa`, `oedit 2s/dfa` ] should show "invalid command format [how to use]"
3. Command with integer <= 0 [ `odel -4` ] should show "Index is not a non-zero unsigned integer" 
4. Command with integer > valid range [ `odel 200` ] should show "The room index provided is invalid"
5. Check that all the edit commands specify that one of the optional params need to be specified.

Please run anything else that might break it or not perform in a user-friendly way to see if it still works.

### CODE REVIEW

**Comment by DrWala** [comment]
Need to make the change for `iclo` too. I have not done that and will do it soon.

**Comment by DrWala** [comment]
> Need to make the change for `iclo`, `alloc`, `dealloc` too. I have not done that and will do it soon.

This has been done in commit 06fed88

**Review by benedictkhoomw — APPROVED** [review]
Smoke tested without issue. 

Tested negative, zero, too large and correct indices for `redit`, `rdel`, `oedit`, `odel`, `iedit`, `idel`, `iclo`, `alloc` and `dealloc`. The messages given had consistent behavior. Thank you for making this.

File: src/main/java/seedu/address/commons/util/StringUtil.java [hunk 1]
```diff
@@ -99,4 +99,21 @@ public static boolean isNonZeroUnsignedInteger(String s) {
             return false;
         }
     }
+    /**
+     * Returns true if {@code s} represents an integer
+     * e.g. -2, -1, 0, 1, 2, 3, ..., {@code Integer.MAX_VALUE} <br>
+     * Will return false for any other non-null string input
+     * e.g. empty string, "abc", "AAA", and " 2 " (untrimmed), "3 0" (contains whitespace), "1 a" (contains letters)
+     *
+     * @throws NullPointerException if {@code s} is null.
```
Comment (stein414) [inline src/main/java/seedu/address/commons/util/StringUtil.java:108]: Javadoc needs param and return too but can be added in a future PR

File: src/test/java/seedu/address/logic/parser/issue/EditIssueCommandParserTest.java [hunk 2]
```diff
@@ -59,16 +60,18 @@ public void parse_missingParts_failure() {
     @Test
     public void parse_invalidPreamble_failure() {
         // negative index
-        assertParseFailure(parser, "-5" + ROOM_NUMBER_DESC_10_100, MESSAGE_INVALID_FORMAT);
+        assertParseFailure(parser, "-5" + ROOM_NUMBER_DESC_10_100, MESSAGE_INVALID_INDEX_RANGE);
 
         // zero index
-        assertParseFailure(parser, "0" + ROOM_NUMBER_DESC_10_100, MESSAGE_INVALID_FORMAT);
+        assertParseFailure(parser, "0" + ROOM_NUMBER_DESC_10_100, MESSAGE_INVALID_INDEX_RANGE);
```
Comment (stein414) [inline src/test/java/seedu/address/logic/parser/issue/EditIssueCommandParserTest.java:66]: Thanks for updating my tests cases too

**Review by stein414 — APPROVED** [review]
LGTM. Thanks for the fix to make it all consistent.

**Review by benedictkhoomw — APPROVED** [review]
Smoke tested updated version without issue.

Tested negative, zero, too large and correct indices for `redit`, `rdel`, `oedit`, `odel`, `iedit`, `idel`, `iclo`, `alloc` and `dealloc` with empty and non-empty lists. The messages given had consistent behavior.

LGTM.

**Review by colintkn — APPROVED** [review]
Tested it. Works great. Big thanks for the fix, Azeem.

**Review by stein414 — APPROVED** [review]
LGTM. Smoked tested the same commands again, seems to be working as expected. Thanks for the update.

**Pull request merged**